geom/twkb_parser.go

```diff
@@ -0,0 +1,562 @@
+	}
+	return p.nextGeometryCollection()
+}
+
+func (p *TWKBParser) nextGeometryCollection() (GeometryCollection, error) {
+	numGeoms, err := p.parseUnsignedVarint()
+	if err != nil {
+		return GeometryCollection{}, p.parserError("cannot unmarshal num polygons varint malformed")
+	}
+	_, err = p.parseIDList(int(numGeoms))
+	if err != nil {
+		return GeometryCollection{}, err
+	}
+	var geoms []Geometry
+	for i := 0; i < int(numGeoms); i++ {
+		subParser := newTWKBParser(p.twkb[p.pos:], p.opts...)
+		geom, nbytes, err := subParser.parseGeometry()
+		if err != nil {
+			return GeometryCollection{}, err
+		}
+		p.pos += nbytes // Sub-parser's geometry has been read, so ensure it is skipped.
+		geoms = append(geoms, geom)
+	}
+	return NewGeometryCollection(geoms, p.opts...), nil
+}
+
+// Read a number of points then convert that many points from int to float coords.
+// Utilise and update the running memory of the previous reference point.
+// Return the slice of coords, the number of points, and any error.
+func (p *TWKBParser) parsePointCountAndArray() ([]float64, int, error) {
+	numPoints, err := p.parseUnsignedVarint()
+	if err != nil {
+		return nil, 0, p.parserError("cannot unmarshal num points varint malformed")
+	}
+
+	coords, err := p.parsePointArray(int(numPoints))
+	return coords, int(numPoints), err
+}
+
+// Convert a given number of points from integer to floating point coordinates.
+// Utilise and update the running memory of the previous reference point.
+// The returned array will contain numPoints * the number of dimensions values.
+func (p *TWKBParser) parsePointArray(numPoints int) ([]float64, error) {
+	var coords = make([]float64, numPoints*p.dimensions)
+	c := 0
+	for i := 0; i < numPoints; i++ {
+		for d := 0; d < p.dimensions; d++ {
+			val, err := p.parseSignedVarint()
+			if err != nil {
+				return nil, p.parserError("cannot unmarshal point %d of %d coord %d varint malformed", i, numPoints, d)
+			}
+
+			p.refpoint[d] += val // Reverse coord differencing to find the true value.
+			coords[c] = float64(p.refpoint[d]) * p.scalings[d]
+			c++
+		}
+	}
+	return coords, nil
+}
+
+func (p *TWKBParser) parseIDList(numIDs int) ([]int, error) {
+	if !p.hasIDs {
+		return nil, nil
+	}
+	var ids = make([]int, numIDs)
+	for i := 0; i < numIDs; i++ {
+		id, err := p.parseSignedVarint()
+		if err != nil {
+			return nil, p.parserError("cannot unmarshal ID varint in ID list")
+		}
+		ids = append(ids, int(id))
+	}
+	return ids, nil
+}
+
+func (p *TWKBParser) parseUnsignedVarint() (uint64, error) {
+	// LEB128.
+	val, n := binary.Uvarint(p.twkb[p.pos:])
+	if n <= 0 {
+		return 0, p.parserError("problem parsing unsigned varint")
+	}
+	p.pos += n // Have now read the varint.
+	return val, nil
+}
+
+func (p *TWKBParser) parseSignedVarint() (int64, error) {
+	// LEB128 Zig-Zag encoded.
+	uval, n := binary.Uvarint(p.twkb[p.pos:])
+	if n <= 0 {
+		return 0, p.parserError("problem parsing signed varint")
```

Comment:
The [docs](https://pkg.go.dev/encoding/binary#Varint) differentiate `n == 0` and `n < 0`, which have different meanings. Do you think it's worth conveying these meanings in separate error messages?

```
n == 0: buf too small
n  < 0: value larger than 64 bits (overflow)
        and -n is the number of bytes read
```

Reply:
I didn't bother differentiating those because they're going to lead to the parse being aborted either way, but you're right, separate error messages could help with debugging.